src/extra/utils/ftd.pyx

```diff
@@ -0,0 +1,532 @@
+
+    # A rounded-up delay is required to control various loops, and also
+    # used instead of the floating delay if its value is actually an
+    # integer. Omitting sinc interpolation in these cases is a huge
+    # performance increase.
+    cdef int int_delay = <int>ceil(delay)
+    cdef bint is_integer_delay = int_delay == delay
+
+    with nogil:
+        for i in range(int_delay, signal.shape[0] - 1):
+            if negative:
+                if signal[i] >= threshold:
+                    continue
+            elif signal[i] <= threshold:
+                continue
+
+            j = i + 1
+
+            if not is_integer_delay:
+                if negative:
+                    cfd_i = fraction * signal[i] - _sinc_interp(
+                        signal, <data_t>i - delay)
+                    cfd_j = fraction * signal[j] - _sinc_interp(
+                        signal, <data_t>j - delay)
+                else:
+                    cfd_i = _sinc_interp(
+                        signal, <data_t>i - delay) - fraction * signal[i]
+                    cfd_j = _sinc_interp(
+                        signal, <data_t>j - delay) - fraction * signal[j]
+            elif negative:
+                cfd_i = fraction * signal[i] - signal[i - int_delay]
+                cfd_j = fraction * signal[j] - signal[j - int_delay]
+            else:
+                cfd_i = signal[i - int_delay] - fraction * signal[i]
+                cfd_j = signal[j - int_delay] - fraction * signal[j]
+
+            # From this point on, the CFD values are computed in such a
+            # way as if the signal and threshold would be positive.
+
+            if cfd_i < walk and cfd_j > walk and i > next_edge:
+                amplitude = signal[i - int_delay]
+
+                if negative:
+                    for k in range(j - int_delay, i + int_delay):
+                        if signal[k] < amplitude:
+                            amplitude = signal[k]
+                else:
+                    for k in range(j - int_delay, i + int_delay):
+                        if signal[k] > amplitude:
+                            amplitude = signal[k]
+
+                if interp == EdgeInterpolation.NEAREST:
+                    edge_pos = i if fabs(cfd_j - walk) > fabs(cfd_i - walk) \
+                        else j
+                elif interp == EdgeInterpolation.LINEAR:
+                    edge_pos = i + (cfd_i - walk) / (cfd_i - cfd_j)
+                elif interp == EdgeInterpolation.SPLINE:
+                    raise NotImplementedError('spline interpolation')
+                elif interp == EdgeInterpolation.SINC:
+                    # As sinc interpolation still makes use of the
+                    # original (possibly negative!) signal, it also
+                    # requires the original walk value for comparions.
+                    edge_pos = _cfd_sinc_interp(signal, i, j, delay, fraction,
+                                                orig_walk)
```

Comment:
As things stand, I think this line will briefly re-acquire the GIL to check for exceptions.

Normally with a numeric return type, Cython returns `-1` to indicate an exception, and after a call only looks for an exception if that's the return value. This is the same as specifying `except? -1`. But it's not smart enough to do this with your specialising `data_t`, and even adding `except? -1` explicitly produces an error.

We could live with that, as this isn't in a hot loop. If you want to get rid of it, the simplest way might be to 'return' the interpolated value via a pointer, and let the actual return value be a simple int so it can indicate failure.

Reply:
I'm not so sure. Checking the annotated output, the GIL seems to only be acquired when the exception is actually raised. When entering the function, only Cython-internel helper functions are called (`__Pyx_RefNannySetupContext`), presumably to prepare for such an event. There's nothing on the calling site.

In any case, that would be acceptable. I would not consider anything using sinc interpolation to be a high performance path - if you're enabling that, you realize it's going to be slower. To give you an idea, it's a 4x penalty to use it for edge position, and another (!) 20x for real delays.

Reply:
Is it possible Cython 3 changed this to handle the exception check at the called site now instead?

Reply:
I think this tidbit should apply in this case and you're partially right:

> There is an easy-to-encounter performance pitfall here with nogil functions with an implicit exception specification of except *. This can happen most commonly when the return type is void (but in principle applies to most non-numeric return types). In this case, Cython is forced to re-acquire the GIL briefly after each call to check the exception state. To avoid this overhead, either change the signature to noexcept (if you have determined that it’s suitable to do so), or to returning an int instead to let Cython use the int as an error flag (by default, -1 triggers the exception check).

There still does not seem to be any GIL acquisition, likely because Cython looks at each type of the `data_t` fused type separately and can use it to propagate exception state. But -1.0 is a valid return value of this function, so let's go with your idea of passing by pointer.